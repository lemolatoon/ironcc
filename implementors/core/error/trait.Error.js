(function() {var implementors = {
"ironcc":[["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.67.0/core/error/trait.Error.html\" title=\"trait core::error::Error\">Error</a> for <a class=\"struct\" href=\"ironcc/error/struct.CompileError.html\" title=\"struct ironcc::error::CompileError\">CompileError</a>"]]
};if (window.register_implementors) {window.register_implementors(implementors);} else {window.pending_implementors = implementors;}})()