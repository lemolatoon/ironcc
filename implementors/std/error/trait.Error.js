(function() {var implementors = {};
implementors["ironcc"] = [{"text":"impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.64.0/std/error/trait.Error.html\" title=\"trait std::error::Error\">Error</a> for <a class=\"struct\" href=\"ironcc/error/struct.CompileError.html\" title=\"struct ironcc::error::CompileError\">CompileError</a>","synthetic":false,"types":["ironcc::error::CompileError"]}];
if (window.register_implementors) {window.register_implementors(implementors);} else {window.pending_implementors = implementors;}})()