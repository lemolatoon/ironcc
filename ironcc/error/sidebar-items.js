window.SIDEBAR_ITEMS = {"enum":[["AnalyzeErrorKind",""],["CompileErrorKind",""],["GenerateErrorKind",""],["NotAllowedStmtKind",""],["ParseErrorKind",""],["TokenizeErrorKind",""],["Tokens",""],["TypeErrorKind",""],["TypeExpectedFailedKind",""],["UnexpectedTypeSizeStatus",""],["VariableKind",""]],"struct":[["CompileError",""]]};