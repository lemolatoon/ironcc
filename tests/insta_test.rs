use insta::assert_debug_snapshot;
pub mod test_utils;
use ironcc::error::CompileError;
use test_utils::CachedProcessor;

macro_rules! all {
    ($self: ident) => {
        assert_debug_snapshot!($self.tokens().as_ref().unwrap());
        assert_debug_snapshot!($self.program().as_ref().unwrap());
        assert_debug_snapshot!($self.conv_program().as_ref().unwrap());
    };
}

#[test]
fn insta_tests() -> Result<(), CompileError> {
    let src = "\nint main() {\nint i;\ni = 5;\nint* p; p = &i;\nint *p2; p2 = p + i;\n}";
    let mut tester = CachedProcessor::new(src);
    all!(tester);
    Ok(())
}

#[test]
fn initializer() -> Result<(), CompileError> {
    let src = "int main() {int a = 5; int *p = &a; return 0;}";
    let mut tester = CachedProcessor::new(src);
    all!(tester);
    Ok(())
}

#[test]
fn size_of() -> Result<(), CompileError> {
    let src = "\n\
    int main() {\n \
        int a = 5;\n \
        int b = sizeof (a);\n \
        int c = sizeof (int);\n \
        int d = sizeof (int*);\n \
    }\n\
    ";
    let mut tester = CachedProcessor::new(src);
    all!(tester);
    Ok(())
}

#[test]
fn array_syntax_sugar() -> Result<(), CompileError> {
    let src = "\n\
    int main() {\n \
        int array[5];
        return array[0];
    }\n\
    ";
    let mut tester = CachedProcessor::new(src);
    all!(tester);
    Ok(())
}

#[test]
fn scopes() {
    let src = "\n\
    int main() {\n \
        int a;
        {int *k;}
        {int *i; {int k;}}
        int c[2];
    }\n\
    ";
    let mut tester = CachedProcessor::new(src);
    all!(tester);
}

#[test]
fn implicit_cast() {
    let src = "
    int main() {
        int a = 2;
        char b = a;
        b = 4;
    }
    ";
    let mut tester = CachedProcessor::new(src);
    all!(tester);
}

#[test]
fn add_char() {
    let src = "
    int main() {
        int a = 2;
        char b = a;
        a + b;
        b + a;
    }
    ";
    let mut tester = CachedProcessor::new(src);
    all!(tester);
}

#[test]
fn string_literal() {
    let src = r#"
    int printf(char *msg);

    int main() {
        char *msg = "Hello World";
        printf(msg);
    }
    "#;
    let mut tester = CachedProcessor::new(src);
    all!(tester);
}

#[test]
fn string_literal2() {
    let src = r#"
    int printf(char *msg);

    int main() {
        char *msg;
        msg = "Hello World";
        printf(msg);
    }
    "#;
    let mut tester = CachedProcessor::new(src);
    all!(tester);
}
