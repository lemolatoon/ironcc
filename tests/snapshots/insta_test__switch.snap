---
source: tests/insta_test.rs
expression: tester.tokens().as_ref().unwrap()
---
[
    Token {
        kind: Type(
            Int,
        ),
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 0,
                n_line: 0,
            },
        },
    },
    Token {
        kind: Ident(
            "f",
        ),
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 4,
                n_line: 0,
            },
        },
    },
    Token {
        kind: OpenDelim(
            Paren,
        ),
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 5,
                n_line: 0,
            },
        },
    },
    Token {
        kind: CloseDelim(
            Paren,
        ),
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 6,
                n_line: 0,
            },
        },
    },
    Token {
        kind: Semi,
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 7,
                n_line: 0,
            },
        },
    },
    Token {
        kind: Type(
            Int,
        ),
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 0,
                n_line: 2,
            },
        },
    },
    Token {
        kind: Ident(
            "main",
        ),
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 4,
                n_line: 2,
            },
        },
    },
    Token {
        kind: OpenDelim(
            Paren,
        ),
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 8,
                n_line: 2,
            },
        },
    },
    Token {
        kind: CloseDelim(
            Paren,
        ),
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 9,
                n_line: 2,
            },
        },
    },
    Token {
        kind: OpenDelim(
            Brace,
        ),
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 11,
                n_line: 2,
            },
        },
    },
    Token {
        kind: Switch,
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 2,
                n_line: 3,
            },
        },
    },
    Token {
        kind: OpenDelim(
            Paren,
        ),
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 9,
                n_line: 3,
            },
        },
    },
    Token {
        kind: Ident(
            "f",
        ),
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 10,
                n_line: 3,
            },
        },
    },
    Token {
        kind: OpenDelim(
            Paren,
        ),
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 11,
                n_line: 3,
            },
        },
    },
    Token {
        kind: CloseDelim(
            Paren,
        ),
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 12,
                n_line: 3,
            },
        },
    },
    Token {
        kind: CloseDelim(
            Paren,
        ),
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 13,
                n_line: 3,
            },
        },
    },
    Token {
        kind: OpenDelim(
            Brace,
        ),
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 15,
                n_line: 3,
            },
        },
    },
    Token {
        kind: Case,
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 4,
                n_line: 4,
            },
        },
    },
    Token {
        kind: Num(
            0,
        ),
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 9,
                n_line: 4,
            },
        },
    },
    Token {
        kind: Colon,
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 10,
                n_line: 4,
            },
        },
    },
    Token {
        kind: Return,
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 6,
                n_line: 5,
            },
        },
    },
    Token {
        kind: Num(
            0,
        ),
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 13,
                n_line: 5,
            },
        },
    },
    Token {
        kind: Semi,
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 14,
                n_line: 5,
            },
        },
    },
    Token {
        kind: Case,
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 4,
                n_line: 6,
            },
        },
    },
    Token {
        kind: Num(
            1,
        ),
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 9,
                n_line: 6,
            },
        },
    },
    Token {
        kind: Colon,
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 10,
                n_line: 6,
            },
        },
    },
    Token {
        kind: Return,
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 6,
                n_line: 7,
            },
        },
    },
    Token {
        kind: Num(
            1,
        ),
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 13,
                n_line: 7,
            },
        },
    },
    Token {
        kind: Semi,
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 14,
                n_line: 7,
            },
        },
    },
    Token {
        kind: Default,
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 4,
                n_line: 8,
            },
        },
    },
    Token {
        kind: Colon,
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 11,
                n_line: 8,
            },
        },
    },
    Token {
        kind: Return,
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 6,
                n_line: 9,
            },
        },
    },
    Token {
        kind: Num(
            666,
        ),
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 13,
                n_line: 9,
            },
        },
    },
    Token {
        kind: Semi,
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 16,
                n_line: 9,
            },
        },
    },
    Token {
        kind: CloseDelim(
            Brace,
        ),
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 2,
                n_line: 10,
            },
        },
    },
    Token {
        kind: CloseDelim(
            Brace,
        ),
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 0,
                n_line: 11,
            },
        },
    },
    Token {
        kind: Eof,
        debug_info: DebugInfo {
            file_info: ,
            pos: Position {
                n_char: 0,
                n_line: 11,
            },
        },
    },
]
